src/Raven.Server/Smuggler/Documents/CsvStreamSource.cs

```diff
@@ -0,0 +1,213 @@
+
+        private DocumentItem ConvertRecordToDocumentItem(string[] csvReaderCurrentRecord, string[] csvReaderFieldHeaders, string collection)
+        {
+            var idStr = _hasId ? csvReaderCurrentRecord[_idIndex] : _hasCollection? $"{csvReaderCurrentRecord[_collectionIndex]}/" : $"{collection}/";
+            var data = new DynamicJsonValue();
+            for (int i = 0; i < csvReaderFieldHeaders.Length; i++)
+            {
+                //ignoring reserved properties
+                if (csvReaderFieldHeaders[i][0] == '@')
+                {
+                    if (_hasCollection && i == _collectionIndex)
+                    {
+                        var metadata = new DynamicJsonValue
+                        {
+                            [nameof(Constants.Documents.Metadata.Collection)] = csvReaderCurrentRecord[_collectionIndex]
+                        };
+                        data[Constants.Documents.Metadata.Key] = metadata;
+                    }
+                    continue;
+                }
+
+                if (_nestedPropertyDictionary != null && _nestedPropertyDictionary.TryGetValue(i, out var segments))
+                {
+                    var nestedData = data;
+                    for (var j = 0;; j++)
+                    {
+                        //last segment holds the data
+                        if (j == segments.Length - 1)
+                        {
+                            nestedData[segments[j]] = ParseValue(csvReaderCurrentRecord[i]);
+                            break; //we are done
+                        }
+                        //Creating the objects along the path if needed e.g. Foo.Bar.Name will create the 'Bar' oject if needed
+                        if (nestedData[segments[j]] == null)
+                        {
+                            var tmpRef = new DynamicJsonValue();
+                            nestedData[segments[j]] = tmpRef;
+                            nestedData = tmpRef;
+                        }
+                    }
+                    continue;
+                }
+                data[csvReaderFieldHeaders[i]] = ParseValue(csvReaderCurrentRecord[i]);
+            }
+
+            return new DocumentItem
+            {
+                Document = new Document
+                {
+                    Data = _context.ReadObject(data,idStr),
+                    Id = _context.GetLazyString(idStr),
+                    ChangeVector = string.Empty,
+                    Flags = DocumentFlags.None,
+                    NonPersistentFlags = NonPersistentDocumentFlags.FromSmuggler
+                },
+                Attachments = null
+            };
+        }
+
+        private unsafe object ParseValue(string s)
+        {
+            if (s.StartsWith('\"') && s.EndsWith('\"'))
+                return s.Substring(1, s.Length-2);
+            if (s.StartsWith('[') && s.EndsWith(']'))
+            {
+                var bytes = Encoding.UTF8.GetBytes(s);
```

Comment:
This allocates a temporary buffer per value, why? reuse an existing buffer

Reply:
Any suggestion would be appreciated

Reply:
done